cl/validator/committee_subscription/committee_subscription.go

```diff
@@ -0,0 +1,348 @@
+		return ErrNotInPropagationRange
+	}
+	// [REJECT] The attestation's epoch matches its target -- i.e. attestation.data.target.epoch == compute_epoch_at_slot(attestation.data.slot)
+	if epoch != slot/c.beaconConfig.SlotsPerEpoch {
+		return ErrEpochMismatch
+	}
+	//[REJECT] The attestation is unaggregated -- that is, it has exactly one participating validator (len([bit for bit in aggregation_bits if bit]) == 1, i.e. exactly 1 bit is set).
+	emptyCount := 0
+	for i := 0; i < len(bits); i++ {
+		if bits[i] == 0 {
+			emptyCount++
+		} else if bits[i]&(bits[i]-1) != 0 {
+			return ErrExactlyOneBitSet
+		}
+	}
+	if emptyCount != len(bits)-1 {
+		return ErrExactlyOneBitSet
+	}
+	// [REJECT] The number of aggregation bits matches the committee size -- i.e. len(aggregation_bits) == len(get_beacon_committee(state, attestation.data.slot, index)).
+	if len(bits)*8 != int(committeeCount) {
+		return ErrAggregationBitsMismatch
+	}
+	// todo ...
+	// [IGNORE] There has been no other valid attestation seen on an attestation subnet that has an identical attestation.data.target.epoch and participating validator index.
+	// [REJECT] The signature of attestation is valid.
+	// [IGNORE] The block being voted for (attestation.data.beacon_block_root) has been seen (via both gossip and non-gossip sources)
+	// (a client MAY queue attestations for processing once block is retrieved).
+	// [REJECT] The block being voted for (attestation.data.beacon_block_root) passes validation.
+	// [REJECT] The attestation's target block is an ancestor of the block named in the LMD vote -- i.e.
+	// get_checkpoint_block(store, attestation.data.beacon_block_root, attestation.data.target.epoch) == attestation.data.target.root
+	// [IGNORE] The current finalized_checkpoint is an ancestor of the block defined by attestation.data.beacon_block_root --
+	// i.e. get_checkpoint_block(store, attestation.data.beacon_block_root, store.finalized_checkpoint.epoch) == store.finalized_checkpoint.root
+	return nil
+}
+
+func (c *CommitteeSubscribeMgmt) OnReceiveAttestation(topic string, att *solid.Attestation) error {
+	if err := c.checkAttestation(topic, att); err != nil {
+		return err
+	}
+
+	var (
+		slot           = att.AttestantionData().Slot()
+		committeeIndex = att.AttestantionData().CommitteeIndex()
+		sig            = att.Signature()
+		bits           = att.AggregationBits()
+		aggrId         = toAggregationId(slot, committeeIndex)
+	)
+	c.aggregationMutex.Lock()
+	defer c.aggregationMutex.Unlock()
+	aggrData, exist := c.aggregations[aggrId]
+	if !exist {
+		// no one is interested in this aggregation
+		return nil
+	}
+	// check if this is the first attestation for this aggregation
+	bitGroupIdx, err := findbitGroup(bits, aggrData.bits)
+	if err == errBitExist {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	// aggregate
+	sigBytes := make([]byte, 96)
+	copy(sigBytes, sig[:])
+	if aggrData != nil {
+		aggrSig, err := bls.AggregateSignatures([][]byte{
```

Comment:
makes sense but you need to aggregate the public keys as well

Reply:
Can I use `BeaconState.ValidatorPublicKey()` to get pub key of specific validator?
And curious when do we use the aggregated public keys?

Reply:
oh good catch, yes public key aggregation is not required for this